market_maker_keeper/api.py

```diff
@@ -0,0 +1,117 @@
+# This file is part of Maker Keeper Framework.
+#
+# Copyright (C) 2020 MikeHathaway
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from argparse import Namespace
+
+from market_maker_keeper.band import Bands
+from market_maker_keeper.control_feed import create_control_feed
+from market_maker_keeper.limit import History
+from market_maker_keeper.order_book import OrderBookManager
+from market_maker_keeper.order_history_reporter import create_order_history_reporter
+from market_maker_keeper.price_feed import PriceFeedFactory
+from market_maker_keeper.reloadable_config import ReloadableConfig
+from market_maker_keeper.spread_feed import create_spread_feed
+from market_maker_keeper.util import setup_logging
+
+from pymaker.lifecycle import Lifecycle
+from pymaker.numeric import Wad
+from pyexchange.api import PyexAPI
+
+
+class KeeperAPI:
+    """
+    Define a common abstract API for keepers
+    """
+
+    def __init__(self, arguments: Namespace, pyex_api: PyexAPI):
+
+        setup_logging(arguments)
+
+        self.bands_config = ReloadableConfig(arguments.config)
+        self.price_feed = PriceFeedFactory().create_price_feed(arguments)
+        self.spread_feed = create_spread_feed(arguments)
+        self.control_feed = create_control_feed(arguments)
+        self.order_history_reporter = create_order_history_reporter(arguments)
+
+        self.history = History()
+
+        self.order_book_manager = OrderBookManager(refresh_frequency=arguments.refresh_frequency)
+        self.order_book_manager.get_orders_with(lambda: pyex_api.get_orders(self.pair()))
+        self.order_book_manager.get_balances_with(lambda: pyex_api.get_balances())
+        self.order_book_manager.cancel_orders_with(lambda order: pyex_api.cancel_order(order.order_id))
+        self.order_book_manager.enable_history_reporting(self.order_history_reporter, self.our_buy_orders,
+                                                         self.our_sell_orders)
+        self.order_book_manager.start()
+
+    def main(self):
+        with Lifecycle() as lifecycle:
+            lifecycle.initial_delay(10)
+            lifecycle.every(1, self.synchronize_orders)
```

Comment:
this should include `on_startup` method as well

Reply:
what would be going into the `on_startup` method?

Reply:
Some keepers are using that to set allowance for traded tokens, e.g. https://github.com/makerdao/market-maker-keeper/blob/master/market_maker_keeper/zrx_market_maker_keeper.py#L205

Reply:
So I started building out an implementation that added that, but noticed that it added a significant amount of complexity that was unneeded for most of the exchanges and DyDx - e.g. had to now add support for pulling web3 out of args, non PyexAPI exchanges like the ZRX ones (which removes the typing protection), and additional token wrangling in token_buy and token_sell. 

Would you be open to leaving this as is, and instead creating a second abstract class that Zrx based keepers would inherit from?

Reply:
:+1: